File: research/experiments/mathexpr-prompt-python-sonnet/mathexpr.py
"""Math expression evaluator with support for basic arithmetic operations."""

from enum import Enum, auto
from typing import Union


class TokenType(Enum):
    """Token types for the lexer."""
    NUMBER = auto()
    PLUS = auto()
    MINUS = auto()
    MULTIPLY = auto()
    DIVIDE = auto()
    MODULO = auto()
    POWER = auto()
    LPAREN = auto()
    RPAREN = auto()
    EOF = auto()


class Token:
    """Represents a token in the expression."""
    
    def __init__(self, type_: TokenType, value: Union[float, str, None] = None):
        self.type = type_
        self.value = value
    
    def __repr__(self):
        return f"Token({self.type}, {self.value})"


class Lexer:
    """Tokenizes a math expression string."""
    
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.current_char = self.text[0] if text else None
    
    def error(self, msg: str = "Invalid character"):
        raise ValueError(f"{msg}: '{self.current_char}' at position {self.pos}")
    
    def advance(self):
        """Move to the next character."""
        self.pos += 1
        if self.pos >= len(self.text):
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]
    
    def skip_whitespace(self):
        """Skip whitespace characters."""
        while self.current_char is not None and self.current_char.isspace():
            self.advance()
    
    def read_number(self) -> float:
        """Read a number (integer or decimal)."""
        num_str = ""
        has_dot = False
        
        while self.current_char is not None and (self.current_char.isdigit() or self.current_char == '.'):
            if self.current_char == '.':
                if has_dot:
                    raise ValueError(f"Invalid number format: multiple decimal points at position {self.pos}")
                has_dot = True
            num_str += self.current_char
            self.advance()
        
        if not num_str or num_str == '.':
            raise ValueError(f"Invalid number format at position {self.pos}")
        
        return float(num_str)
    
    def get_next_token(self) -> Token:
        """Get the next token from the input."""
        while self.current_char is not None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue
            
            if self.current_char.isdigit() or self.current_char == '.':
                return Token(TokenType.NUMBER, self.read_number())
            
            if self.current_char == '+':
                self.advance()
                return Token(TokenType.PLUS)
            
            if self.current_char == '-':
                self.advance()
                return Token(TokenType.MINUS)
            
            if self.current_char == '*':
                self.advance()
                # Check for ** (power operator)
                if self.current_char == '*':
                    self.advance()
                    return Token(TokenType.POWER)
                return Token(TokenType.MULTIPLY)
            
            if self.current_char == '/':
                self.advance()
                return Token(TokenType.DIVIDE)
            
            if self.current_char == '%':
                self.advance()
                return Token(TokenType.MODULO)
            
            if self.current_char == '(':
                self.advance()
                return Token(TokenType.LPAREN)
            
            if self.current_char == ')':
                self.advance()
                return Token(TokenType.RPAREN)
            
            self.error("Invalid character")
        
        return Token(TokenType.EOF)


class Parser:
    """Parses tokens into an abstract syntax tree."""
    
    def __init__(self, lexer: Lexer):
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token()
    
    def error(self, msg: str = "Invalid syntax"):
        raise ValueError(msg)
    
    def eat(self, token_type: TokenType):
        """Consume a token of the given type."""
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error(f"Expected {token_type}, got {self.current_token.type}")
    
    def parse(self) -> float:
        """Parse and evaluate the expression."""
        if self.current_token.type == TokenType.EOF:
            raise ValueError("Empty input")
        
        result = self.expression()
        
        if self.current_token.type != TokenType.EOF:
            if self.current_token.type == TokenType.RPAREN:
                raise ValueError("Unmatched parentheses: unexpected closing parenthesis")
            self.error(f"Unexpected token: {self.current_token.type}")
        
        return result
    
    def expression(self) -> float:
        """Handle addition and subtraction (lowest precedence)."""
        result = self.term()
        
        while self.current_token.type in (TokenType.PLUS, TokenType.MINUS):
            token = self.current_token
            if token.type == TokenType.PLUS:
                self.eat(TokenType.PLUS)
                result = result + self.term()
            elif token.type == TokenType.MINUS:
                self.eat(TokenType.MINUS)
                result = result - self.term()
        
        return result
    
    def term(self) -> float:
        """Handle multiplication, division, and modulo."""
        result = self.unary()
        
        while self.current_token.type in (TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MODULO):
            token = self.current_token
            if token.type == TokenType.MULTIPLY:
                self.eat(TokenType.MULTIPLY)
                result = result * self.unary()
            elif token.type == TokenType.DIVIDE:
                self.eat(TokenType.DIVIDE)
                divisor = self.unary()
                if divisor == 0:
                    raise ValueError("Division by zero")
                result = result / divisor
            elif token.type == TokenType.MODULO:
                self.eat(TokenType.MODULO)
                divisor = self.unary()
                if divisor == 0:
                    raise ValueError("Modulo by zero")
                result = result % divisor
        
        return result
    
    def unary(self) -> float:
        """Handle unary minus."""
        token = self.current_token
        
        if token.type == TokenType.MINUS:
            self.eat(TokenType.MINUS)
            return -self.unary()
        
        return self.factor()
    
    def factor(self) -> float:
        """Handle exponentiation (right-associative)."""
        result = self.primary()
        
        if self.current_token.type == TokenType.POWER:
            self.eat(TokenType.POWER)
            # Right-associative: recursively parse the right side
            result = result ** self.unary()
        
        return result
    
    def primary(self) -> float:
        """Handle numbers and parentheses."""
        token = self.current_token
        
        if token.type == TokenType.NUMBER:
            self.eat(TokenType.NUMBER)
            return token.value
        
        if token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            result = self.expression()
            if self.current_token.type != TokenType.RPAREN:
                raise ValueError("Unmatched parentheses: missing closing parenthesis")
            self.eat(TokenType.RPAREN)
            return result
        
        # Check for common malformed expression patterns
        if token.type in (TokenType.PLUS, TokenType.MULTIPLY, TokenType.DIVIDE, 
                         TokenType.MODULO, TokenType.POWER):
            raise ValueError(f"Malformed expression: unexpected operator {token.type}")
        
        if token.type == TokenType.RPAREN:
            raise ValueError("Unmatched parentheses: unexpected closing parenthesis")
        
        if token.type == TokenType.EOF:
            raise ValueError("Malformed expression: unexpected end of input")
        
        self.error(f"Unexpected token: {token.type}")


def calc(expression: str) -> float:
    """
    Evaluate a math expression and return the result.
    
    Args:
        expression: A string containing a math expression
        
    Returns:
        The numeric result of evaluating the expression
        
    Raises:
        ValueError: If the expression is invalid or cannot be evaluated
    """
    if not expression or not expression.strip():
        raise ValueError("Empty input")
    
    lexer = Lexer(expression)
    parser = Parser(lexer)
    return parser.parse()

File: research/experiments/mathexpr-prompt-go-sonnet/mathexpr.go
package mathexpr

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode"
)

// TokenType represents the type of a token
type TokenType int

const (
	TokenNumber TokenType = iota
	TokenPlus
	TokenMinus
	TokenStar
	TokenSlash
	TokenPercent
	TokenPower
	TokenLeftParen
	TokenRightParen
	TokenEOF
)

// Token represents a lexical token
type Token struct {
	Type  TokenType
	Value string
}

// Tokenizer converts a string into a sequence of tokens
type Tokenizer struct {
	input string
	pos   int
}

// NewTokenizer creates a new tokenizer
func NewTokenizer(input string) *Tokenizer {
	return &Tokenizer{input: input, pos: 0}
}

// peek returns the current character without consuming it
func (t *Tokenizer) peek() rune {
	if t.pos >= len(t.input) {
		return 0
	}
	return rune(t.input[t.pos])
}

// advance moves to the next character
func (t *Tokenizer) advance() rune {
	if t.pos >= len(t.input) {
		return 0
	}
	ch := rune(t.input[t.pos])
	t.pos++
	return ch
}

// skipWhitespace skips whitespace characters
func (t *Tokenizer) skipWhitespace() {
	for t.peek() != 0 && unicode.IsSpace(t.peek()) {
		t.advance()
	}
}

// readNumber reads a number token
func (t *Tokenizer) readNumber() string {
	start := t.pos
	for t.peek() != 0 && (unicode.IsDigit(t.peek()) || t.peek() == '.') {
		t.advance()
	}
	return t.input[start:t.pos]
}

// NextToken returns the next token
func (t *Tokenizer) NextToken() (Token, error) {
	t.skipWhitespace()

	ch := t.peek()
	if ch == 0 {
		return Token{Type: TokenEOF}, nil
	}

	// Numbers (including decimals starting with .)
	if unicode.IsDigit(ch) || ch == '.' {
		num := t.readNumber()
		return Token{Type: TokenNumber, Value: num}, nil
	}

	// Operators
	switch ch {
	case '+':
		t.advance()
		return Token{Type: TokenPlus}, nil
	case '-':
		t.advance()
		return Token{Type: TokenMinus}, nil
	case '*':
		t.advance()
		// Check for **
		if t.peek() == '*' {
			t.advance()
			return Token{Type: TokenPower}, nil
		}
		return Token{Type: TokenStar}, nil
	case '/':
		t.advance()
		return Token{Type: TokenSlash}, nil
	case '%':
		t.advance()
		return Token{Type: TokenPercent}, nil
	case '(':
		t.advance()
		return Token{Type: TokenLeftParen}, nil
	case ')':
		t.advance()
		return Token{Type: TokenRightParen}, nil
	default:
		return Token{}, fmt.Errorf("invalid character: %c", ch)
	}
}

// Tokenize converts the entire input into a list of tokens
func Tokenize(input string) ([]Token, error) {
	tokenizer := NewTokenizer(input)
	var tokens []Token

	for {
		token, err := tokenizer.NextToken()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, token)
		if token.Type == TokenEOF {
			break
		}
	}

	return tokens, nil
}

// Node represents an AST node
type Node interface {
	Evaluate() (float64, error)
}

// NumberNode represents a number literal
type NumberNode struct {
	Value float64
}

func (n *NumberNode) Evaluate() (float64, error) {
	return n.Value, nil
}

// UnaryNode represents a unary operation
type UnaryNode struct {
	Operator TokenType
	Operand  Node
}

func (n *UnaryNode) Evaluate() (float64, error) {
	val, err := n.Operand.Evaluate()
	if err != nil {
		return 0, err
	}

	switch n.Operator {
	case TokenMinus:
		return -val, nil
	default:
		return 0, fmt.Errorf("unknown unary operator")
	}
}

// BinaryNode represents a binary operation
type BinaryNode struct {
	Operator TokenType
	Left     Node
	Right    Node
}

func (n *BinaryNode) Evaluate() (float64, error) {
	left, err := n.Left.Evaluate()
	if err != nil {
		return 0, err
	}

	right, err := n.Right.Evaluate()
	if err != nil {
		return 0, err
	}

	switch n.Operator {
	case TokenPlus:
		return left + right, nil
	case TokenMinus:
		return left - right, nil
	case TokenStar:
		return left * right, nil
	case TokenSlash:
		if right == 0 {
			return 0, fmt.Errorf("division by zero")
		}
		return left / right, nil
	case TokenPercent:
		if right == 0 {
			return 0, fmt.Errorf("modulo by zero")
		}
		return math.Mod(left, right), nil
	case TokenPower:
		return math.Pow(left, right), nil
	default:
		return 0, fmt.Errorf("unknown binary operator")
	}
}

// Parser converts tokens into an AST
type Parser struct {
	tokens []Token
	pos    int
}

// NewParser creates a new parser
func NewParser(tokens []Token) *Parser {
	return &Parser{tokens: tokens, pos: 0}
}

// peek returns the current token without consuming it
func (p *Parser) peek() Token {
	if p.pos >= len(p.tokens) {
		return Token{Type: TokenEOF}
	}
	return p.tokens[p.pos]
}

// advance moves to the next token
func (p *Parser) advance() Token {
	if p.pos >= len(p.tokens) {
		return Token{Type: TokenEOF}
	}
	token := p.tokens[p.pos]
	p.pos++
	return token
}

// expect consumes a token of the given type
func (p *Parser) expect(tokenType TokenType) error {
	token := p.peek()
	if token.Type != tokenType {
		return fmt.Errorf("expected token type %d, got %d", tokenType, token.Type)
	}
	p.advance()
	return nil
}

// Parse parses the tokens into an AST
func (p *Parser) Parse() (Node, error) {
	if p.peek().Type == TokenEOF {
		return nil, fmt.Errorf("empty input")
	}

	node, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	if p.peek().Type != TokenEOF {
		return nil, fmt.Errorf("unexpected token after expression")
	}

	return node, nil
}

// parseExpression parses an expression (lowest precedence: addition/subtraction)
func (p *Parser) parseExpression() (Node, error) {
	return p.parseBinaryOp(p.parseTerm, TokenPlus, TokenMinus)
}

// parseTerm parses a term (multiplication/division/modulo)
func (p *Parser) parseTerm() (Node, error) {
	return p.parseBinaryOp(p.parseExponentiation, TokenStar, TokenSlash, TokenPercent)
}

// parseExponentiation parses exponentiation (right-associative)
func (p *Parser) parseExponentiation() (Node, error) {
	left, err := p.parseUnary()
	if err != nil {
		return nil, err
	}

	// Right-associative: 2 ** 3 ** 2 = 2 ** (3 ** 2)
	if p.peek().Type == TokenPower {
		op := p.advance()
		right, err := p.parseExponentiation() // Recursive call for right associativity
		if err != nil {
			return nil, err
		}
		return &BinaryNode{Operator: op.Type, Left: left, Right: right}, nil
	}

	return left, nil
}

// parseUnary parses a unary expression
func (p *Parser) parseUnary() (Node, error) {
	token := p.peek()

	if token.Type == TokenMinus {
		p.advance()
		operand, err := p.parseUnary() // Allow chaining of unary operators
		if err != nil {
			return nil, err
		}
		return &UnaryNode{Operator: TokenMinus, Operand: operand}, nil
	}

	return p.parsePrimary()
}

// parsePrimary parses a primary expression (number or parenthesized expression)
func (p *Parser) parsePrimary() (Node, error) {
	token := p.peek()

	switch token.Type {
	case TokenNumber:
		p.advance()
		value, err := strconv.ParseFloat(token.Value, 64)
		if err != nil {
			return nil, fmt.Errorf("invalid number: %s", token.Value)
		}
		return &NumberNode{Value: value}, nil

	case TokenLeftParen:
		p.advance()
		node, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		if err := p.expect(TokenRightParen); err != nil {
			return nil, fmt.Errorf("unmatched parentheses")
		}
		return node, nil

	default:
		return nil, fmt.Errorf("malformed expression")
	}
}

// parseBinaryOp is a helper for parsing left-associative binary operators
func (p *Parser) parseBinaryOp(parseNext func() (Node, error), operators ...TokenType) (Node, error) {
	left, err := parseNext()
	if err != nil {
		return nil, err
	}

	for {
		token := p.peek()
		matched := false
		for _, op := range operators {
			if token.Type == op {
				matched = true
				break
			}
		}

		if !matched {
			break
		}

		p.advance()
		right, err := parseNext()
		if err != nil {
			return nil, err
		}

		left = &BinaryNode{Operator: token.Type, Left: left, Right: right}
	}

	return left, nil
}

// Calc evaluates a math expression and returns the result
func Calc(expression string) (float64, error) {
	// Trim whitespace
	expression = strings.TrimSpace(expression)

	// Tokenize
	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}

	// Parse
	parser := NewParser(tokens)
	ast, err := parser.Parse()
	if err != nil {
		return 0, err
	}

	// Evaluate
	return ast.Evaluate()
}
